File: 0739-daily-temperatures/0739-daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        
        stack = []      # items are (index, temperature) pairs
        output = [0] * len(temperatures)    # init with 0s

        # loop thru temperatures by index
        for i, t in enumerate(temperatures):
            # we got a new temp
            # for this t, let's see 
            # how many are waiting in the stack which
            # are < t
            # stack elements are tuple of (index, temp)
            while stack != [] and stack[-1][1] < t:
                # pop the 1st less warm entry
                index, temp = stack.pop()
                dayCount = i - index
                output[index] = dayCount

            # When it's done, push in the current day
            stack.append((i, t))

        return output

File: 0739-daily-temperatures/0739-daily-temperatures.js
/**
 * @param {number[]} temperatures
 * @return {number[]}
 */
var dailyTemperatures = function(temperatures) {
    let ans = new Array(temperatures.length).fill(0);

    // we'll create a monotonic stack
    // and instead of values we will use the index of the values
    // so that when needed we can get both the index and the value
    let stack = [];
    for(let i=0; i<temperatures.length; i++) {
        // if the stack is empty
        // OR is the current temp is less than the stack-top temp
        // just push it
        if(!stack.length || temperatures[i] <= temperatures[stack.at(-1)]) {
            stack.push(i)
        }
        // else, we have found a bigger value
        else {
            // while current temp is > stack-top temp
            while(temperatures[i] > temperatures[stack.at(-1)]) {
                let j = stack.pop();
                ans[j] = i-j;
            }
            // at the end, 
            // don't forget to push the current el to the stack
            stack.push(i)
        }
    }

    return ans;
};